File: ts/src/in-memory-flow-store.ts
import type { FlowContext } from './flow-context.js';
import type { FlowInstance } from './flow-instance.js';

export interface TransitionRecord {
  flowId: string;
  from: string | null;
  to: string;
  trigger: string;
  subFlow: string | null;
  timestamp: Date;
}

export class InMemoryFlowStore {
  private flows = new Map<string, FlowInstance<any>>();
  private _transitionLog: TransitionRecord[] = [];

  /** Clear all flows and transition log. For pool/reuse patterns. */
  clear(): void {
    this.flows.clear();
    this._transitionLog = [];
  }

  create(flow: FlowInstance<any>): void {
    this.flows.set(flow.id, flow);
  }

  loadForUpdate<S extends string>(flowId: string, _definition?: any): FlowInstance<S> | undefined {
    const flow = this.flows.get(flowId);
    if (!flow || flow.isCompleted) return undefined;
    return flow as FlowInstance<S>;
  }

  save(flow: FlowInstance<any>): void {
    this.flows.set(flow.id, flow);
  }

  recordTransition(
    flowId: string, from: string | null, to: string, trigger: string, _ctx: FlowContext,
  ): void {
    const subFlow = trigger.startsWith('subFlow:') ? trigger.substring(8, trigger.indexOf('/')) : null;
    this._transitionLog.push({ flowId, from, to, trigger, subFlow, timestamp: new Date() });
  }

  get transitionLog(): readonly TransitionRecord[] {
    return this._transitionLog;
  }
}

File: java/src/main/java/com/tramli/InMemoryFlowStore.java
package com.tramli;

import java.time.Instant;
import java.util.*;

/**
 * In-memory FlowStore for testing. No external dependencies.
 */
public final class InMemoryFlowStore implements FlowStore {
    private final Map<String, FlowInstance<?>> flows = new LinkedHashMap<>();
    private final List<TransitionRecord> transitionLog = new ArrayList<>();

    public record TransitionRecord(String flowId, String from, String to, String trigger, String subFlow, Instant timestamp) {}

    @Override
    public void create(FlowInstance<?> flow) {
        flows.put(flow.id(), flow);
    }

    @Override
    @SuppressWarnings("unchecked")
    public <S extends Enum<S> & FlowState> Optional<FlowInstance<S>> loadForUpdate(
            String flowId, FlowDefinition<S> definition) {
        FlowInstance<?> flow = flows.get(flowId);
        if (flow == null || flow.isCompleted()) return Optional.empty();
        return Optional.of((FlowInstance<S>) flow);
    }

    @Override
    public void save(FlowInstance<?> flow) {
        flows.put(flow.id(), flow);
    }

    @Override
    public void recordTransition(String flowId, FlowState from, FlowState to,
                                 String trigger, FlowContext ctx) {
        // Extract subFlow name from trigger like "subFlow:payment/DONE"
        String subFlowName = trigger.startsWith("subFlow:") ? trigger.substring(8, trigger.indexOf('/')) : null;
        transitionLog.add(new TransitionRecord(flowId,
                from != null ? from.name() : null, to.name(), trigger, subFlowName, Instant.now()));
    }

    /** Clear all flows and transition log. For pool/reuse patterns. */
    public void clear() {
        flows.clear();
        transitionLog.clear();
    }

    /** Load a flow by ID regardless of completion status (read-only access). */
    @SuppressWarnings("unchecked")
    public <S extends Enum<S> & FlowState> Optional<FlowInstance<S>> load(String flowId) {
        FlowInstance<?> flow = flows.get(flowId);
        if (flow == null) return Optional.empty();
        return Optional.of((FlowInstance<S>) flow);
    }

    public List<TransitionRecord> transitionLog() {
        return Collections.unmodifiableList(transitionLog);
    }
}
